core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala

```diff
@@ -234,19 +235,32 @@ object Protocol {
   }
 
   /**
-   * Extracts all table features that are enabled by the given metadata.
+   * Extracts all table features that are enabled by the given protocol and metadata.
```

Comment:
```suggestion
   * Extracts all table features that are enabled by the given protocol and metadata.
   * This includes features enabled directly by metadata, and their dependencies.
```

Reply:
It would be cleaner if we split this into several methods:
```scala
// give a set of features, return with the set plus all dependencies
def withDependencies(features: Set(TableFeature))

// calls `withDependencies` internally, returns features with dependencies
def extractAutomaticallyEnabledFeatures(spark: SparkSession, metadata: Metadata)

// extractRequiredFeatures(spark, Protocol(), metadata) becomes:
withDependencies(protocol.implicitlyAndExplicitlySupportedFeatures) ++ extractAutomaticallyEnabledFeatures(spark, metadata)
```